### PR TITLE
enh: Adding annotation and UrlRewrites property to configure istio routes

### LINKED ISSUE #3470: Seldon + Istio: Custom URI re-write in VirtualService configuration
I have deployed seldon-core with Istio enabled. i.e. Istio Ingress gateway is used to route the model invoking request to seldon. Model invocation works fine if I hit the endpoint:-
```
https://<hostname>/seldon/<namespace>/<deployment>/api/v1.0/predictions
```

I need the ability to invoke models without the trailing path "/api/v1.0/predictions" i.e. like this:-
```
https://<hostname>/seldon/<namespace>/<deployment>/
```

The virtual service created by seldon controller is as follows:-
```yaml
spec:
  gateways:
  - mxe-senthil/gateway-mxe-csr
  hosts:
  - '*'
  http:
  - match:
    - uri:
        prefix: /seldon/<namepsace>/<deployment>/
    rewrite:
      uri: /
    route:
    - destination:
        host: single-stateless-manual-main
        port:
          number: 8000
        subset: main
```

Request to allow user to configure re-write rules when deploying seldon, so that it creates a virtual service supporting custom uri re-write:-
```yaml
spec:
  gateways:
  - mxe-senthil/gateway-mxe-csr
  hosts:
  - '*'
  http:
  - match:
    - uri:
        exact: /model-endpoints/single-stateless-manual-main
    rewrite:
      uri: /api/v1.0/predictions
    route:
    - destination:
        host: single-stateless-manual-main
        port:
          number: 8000
        subset: main
```

### PR DESCRIPTION
**Pull request by brightsparc**
**What this PR does / why we need it**:
Enables providing custom configuration to may url prefix to rewrite configuration for istio [VirtualService](https://istio.io/latest/docs/reference/config/networking/virtual-service/).

In particular it allows creating the VirtualSerivce with a list of prefixes`/model-endpoints/v2/model-one` that maps to an endpoints `/v2/model-1` which might not have the same model name eg:

```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: seldon-predictor
spec:
  gateways:
  -istio-system/seldon-gateway
  hosts:
  - "*"
  http:
  - route:
    - destination:
        host: seldon-predictor.canary.svc.cluster.local
        port:
          number: 8000
    match:
    - uri:
        prefix: /model-endpoints/v2/model-one
    rewrite:
      uri: /v2/model-1
```

**Which issue(s) this PR fixes**:
Fixes #3470 

**Special notes for your reviewer**:
This is an example of SeldonDeployment configuration that includes the new annotations as well as optional `urlRewrites` dictionary to map specific routes inside the container:

```
apiVersion: machinelearning.seldon.io/v1
kind: SeldonDeployment
metadata:
  name: example-seldon
  annotations:
    "seldon.io/istio-host": "seldon.io"
    "seldon.io/istio-prefix": "/model-endpoints/v2"
    "seldon.io/istio-rewrite": "/v2"
spec:
  protocol: seldon
  urlRewrites:
    "/model-endpoints/v2/model-one": "/v2/model-1"
    "/model-endpoints/v2/model-two": "/v2/model-2"
  predictors:
  - componentSpecs:
    - spec:
        containers:
        - image: seldonio/mock_classifier:{VERSION}
          name: classifier
    graph:
      name: classifier
      type: MODEL
    name: model
    replicas: 1
```

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @brightsparc. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) or [todo](https://github.com/orgs/todo/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/todo/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **majolo**
You can assign the PR to them by writing `/assign @majolo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: operator/controllers/seldondeployment_controller.go
Comment (adriangonz) [inline operator/controllers/seldondeployment_controller.go:238]: Since the default route also gets added through the `UrlRewrites` field - which is part of the editable spec - wouldn't this allow a user to completely remove the default route for Seldon Core? 

These routes are considered part of the standard set of protocols supported by SC, therefore I think they should always be present (even if just on top of custom rewrites).

**Review by adriangonz — CHANGES_REQUESTED** [review]
Hey @brightsparc, 

Thanks for the contribution and apologies for the long delay to review it.

This is quite a big change in behaviour. I've added one of my concerns about the current proposed design in a comment below. It would be good to get your thoughts on that one though.

Besides that point, it would also be good to have both thorough tests and also docs that cover the change. @brightsparc would you be able to add those to the PR?

PS: For future contributions, in cases like this one - where a big change is proposed -  it's generally best to first agree on a design through a regular issue, before contributing any changes.

**Comment by cliveseldon** [comment]
We plan to close this unless you have time to review further @brightsparc

**Comment by brightsparc** [comment]
> We plan to close this unless you have time to review further @brightsparc

I have worked around this further up the stack, fine to close.

**Comment by cliveseldon** [comment]
Thanks @brightsparc

**Pull request closed without merging**
